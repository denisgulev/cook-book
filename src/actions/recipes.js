import uuid from "uuid";
import db from "../firebase/firebase";

/* WHITOUT DB */
/**
 * component calls action generator
 * action generator returns object
 * component dispatches object
 * redux store changes
 */

/* WITH DB */
/**
 * components calls action generator
 * action generator returns function
 * component dispatches function (?)
 * function runs (has the ability to dispatch other actions)
 */

// add recipe

export const addRecipe = recipe => ({
  type: "ADD_RECIPE",
  recipe
});

export const startAddRecipe = (recipeData = {}) => {
  return dispatch => {
    const {
      title = "",
      description = "",
      note = "",
      createdAt = 1000
    } = recipeData;

    const recipe = { title, description, note, createdAt };

    return db
      .ref("recipes")
      .push(recipe)
      .then(ref => {
        dispatch(
          addRecipe({
            id: ref.key,
            ...recipe
          })
        );
      });
  };
};

// remove recipe

export const removeRecipe = ({ id } = {}) => ({
  type: "REMOVE_RECIPE",
  id
});

// edit recipe

export const editRecipe = (id, updates) => ({
  type: "EDIT_RECIPE",
  id,
  updates
});
