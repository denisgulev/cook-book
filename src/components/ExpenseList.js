import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import getVisibleExpenses from '../selectors/expenses';

const ExpenseList = (props) => (
	<div>
		<h1>Expense List</h1>
		{props.expenses.map((expense) => {
			return <ExpenseListItem key={expense.id} {...expense} />
		})}
	</div>
);

const mapStateToProps = (state) => {
	// specify what info from the store we want
	// our component to be able to access
	return {
		expenses: getVisibleExpenses(state.expenses, state.filters)
	};
}

export default connect(mapStateToProps)(ExpenseList);