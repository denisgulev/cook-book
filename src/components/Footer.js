import React from "react";

export const Footer = () => (
    <footer className="footer">
        <div className="footer__content">
            <span>&copy; 2020 - Denis Gulev</span>
        </div>
    </footer>
);

export default Footer;