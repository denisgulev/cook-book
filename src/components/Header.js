import React from "react";
import { Link } from "react-router-dom";
import { connect } from "react-redux";

import RecipeListFilters from "./RecipeListFilters";

export const Header = () => (
  <header className="header">
    <div className="content-container">
      <div className="header__content">
        <Link className="header__title" to="/">
          <h1>CookBook</h1>
        </Link>
        <RecipeListFilters />
      </div>
    </div>
  </header>
);

export default connect()(Header);
