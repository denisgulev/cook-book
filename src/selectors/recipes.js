import moment from "moment";

// get visibleRecipes - 2nd argument is "filterReducerState"
const getVisibleRecipes = (recipes, { text, sortBy, startDate, endDate, category }) => {
  return recipes
    .filter(recipe => {
      const textMatch = recipe.title.toLowerCase().includes(text.toLowerCase());
      if (sortBy === "date") {
        const createdAtMoment = moment(recipe.createdAt);
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, "day") : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, "day") : true;

        return startDateMatch && endDateMatch && textMatch;
      } else if (sortBy === "category") {
        if (recipe.category !== undefined) {
          if (category.toLowerCase() === "all") {
            // show all recipes
            return true && textMatch;
          } else {
            // show recipe with category = searched category
            const categoryMatch = recipe.category.toLowerCase() === category.toLowerCase();

            return categoryMatch && textMatch;
          }
        } else {
          return false;
        }
      }
    })
    .sort((a, b) => {
      if (sortBy === "date") return a.createdAt < b.createdAt ? 1 : -1;
      else if (sortBy === "category") return a.title < b.title ? -1 : 1;
    });
};

export default getVisibleRecipes;
