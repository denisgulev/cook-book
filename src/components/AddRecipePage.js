import React from "react";
import RecipeForm from "./RecipeForm";
import { connect } from "react-redux";
import { startAddRecipe } from "../actions/recipes";

export class AddRecipePage extends React.Component {
  onSubmit = recipe => {
    this.props.startAddRecipe(recipe);
    this.props.history.push("/");
  };

  render() {
    return (
      <div>
        <h1>Add Recipe</h1>
        <RecipeForm onSubmit={this.onSubmit} />
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  startAddRecipe: recipe => dispatch(startAddRecipe(recipe))
});

export default connect(undefined, mapDispatchToProps)(AddRecipePage);
