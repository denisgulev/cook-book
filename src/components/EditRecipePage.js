import React from "react";
import { connect } from "react-redux";
import RecipeForm from "./RecipeForm";
import { startEditRecipe } from "../actions/recipes";

export class EditRecipePage extends React.Component {
  onSubmit = recipe => {
    this.props.startEditRecipe(this.props.recipe.id, recipe);
    this.props.history.push("/dashboard");
  };

  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Edit Recipe</h1>
          </div>
        </div>
        <div className="content-container recipe">
          <RecipeForm recipe={this.props.recipe} onSubmit={this.onSubmit} />
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => ({
  recipe: state.recipes.find(recipe => recipe.id === props.match.params.id)
});

const mapDispatchToProps = (dispatch, props) => ({
  startEditRecipe: (id, recipe) => dispatch(startEditRecipe(id, recipe))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditRecipePage);
