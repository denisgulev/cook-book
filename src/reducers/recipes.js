import * as actions from "../actions/actionTypes";

const recipesReducerDefaultState = [];

const recipesReducer = (state = recipesReducerDefaultState, action) => {
    switch (action.type) {
        case actions.ADD_RECIPE:
            return [...state, action.recipe];
        case actions.REMOVE_RECIPE:
            return state.filter(({id}) => id !== action.id);
        case actions.EDIT_RECIPE:
            return state.map(recipe => {
                if (recipe.id === action.id) {
                    return {
                        ...recipe,
                        ...action.updates
                    };
                } else {
                    return recipe;
                }
            });
        case actions.SET_RECIPES:
            return action.recipes;
        default:
            return state;
    }
};

export default recipesReducer;
