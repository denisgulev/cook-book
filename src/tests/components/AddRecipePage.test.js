import React from "react";
import { shallow } from "enzyme";
import { AddRecipePage } from "../../components/AddRecipePage";
import recipes from "../fixtures/recipes";

let startAddRecipe, history, wrapper;

beforeEach(() => {
  startAddRecipe = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(
    <AddRecipePage startAddRecipe={startAddRecipe} history={history} />
  );
});

test("should render AddRecipePage correctly", () => {
  expect(wrapper).toMatchSnapshot();
});

test("should handle onSubmit", () => {
  wrapper.find("RecipeForm").prop("onSubmit")(recipes[1]);
  expect(history.push).toHaveBeenLastCalledWith("/");
  expect(startAddRecipe).toHaveBeenLastCalledWith(recipes[1]);
});
