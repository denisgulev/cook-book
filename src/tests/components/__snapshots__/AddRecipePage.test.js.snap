// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render AddRecipePage correctly 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Add Recipe
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <RecipeForm
      onSubmit={[Function]}
    />
  </div>
</div>
`;
